R/expect-condition.R

```diff
@@ -106,6 +110,25 @@ expect_condition <- function(object,
   invisible(act$val %||% act$cap)
 }
 
+capture_first_condition <- function(expr) {
+  first <- NULL
+
+  tryCatch(
+    withCallingHandlers(expr, condition = function(cnd) {
+      if (is.null(first)) {
+        if (can_entrace(cnd)) {
+          cnd <- cnd_entrace(cnd)
+        }
+        first <<- cnd
+      }
+      cnd_muffle(cnd)
```

Comment:
I thought `cnd_muffle()` conditionally muffled already 😞 

But we have to muffle messages and warnings, otherwise they bubble up.

Reply:
That makes sense. Though only the first time, right? Ulterior messages and warnings shouldn't be muffled, just like ulterior errors shouldn't be caught.

Reply:
Oh yeah, that makes sense!

Reply:
But really should we capture the first condition in these expectations? Or only the first one that matches the expectation?

If we only capture the condition that matches the expectation, we'd solve #1052 right?

Reply:
It doesn't really help with that issue because that's a totally different code path.

I think we should assume that most of the time when you're using this function there will only be a single condition. I think if I was design this today, I'd just error if there were multiple conditions, but I think that's likely to break existing code.

Reply:
I was suggesting it would make sense to make it use that code path.

I don't think making it fail when there are multiple conditions would be helpful. You'd get failures from causes that are potentially not in your control. I think it should just fail if the expected condition is not there. Similarly, `expect_warning()` should not fail if there are other warnings (they should bubble through), it should only fail if the expected warning isn't there. It isn't always the case right now, which I think is a bug.

Reply:
To do that we'd need to make the implementation of the condition capture considerably more complex, and it just doesn't feel worth the effort to me. In most cases, only one condition is thrown. If there are multiple conditions, I think you're better of capturing them all yourself and using whatever logic makes sense for your specific test case. I agree that your approach is more correct, I just don't think it's worth implementing right now.

Reply:
I'm not sure it would be considerably more complex. We'd just pass "message" and "class" to the capture function and then record the condition only if it matches. Otherwise we let it go unhandled. (Note that as mentioned in my other message I think there is no need for a counter, maybe that was part of the implementation complexity you were thinking about?)

If #1052 is not fixed, I can no longer use it confidently. So I feel like the trade off is worth it.

Reply:
See https://github.com/r-lib/testthat/issues/1052#issuecomment-652797808 for an example that will be rather common and out of the control of the package author.

Reply:
Got it. New coming incoming momentarily 😄